flow360/component/geometry.py

```diff
@@ -186,28 +135,84 @@ def submit(self, progress_callback=None) -> Geometry:
         info = GeometryMeta(**resp)
-        self._id = info.id
-        submitted_geometry = Geometry(self.id)
 
-        remote_file_names = []
-        for geometry_file in self.file_names:
-            remote_file_name = os.path.basename(geometry_file)
-            file_name_to_upload = geometry_file
-            submitted_geometry._upload_file(
-                remote_file_name, file_name_to_upload, progress_callback=progress_callback
+        ##:: upload geometry files
+        geometry = Geometry(info.id)
+        heartbeat_info = {"resourceId": info.id, "resourceType": "Geometry", "stop": False}
+        heartbeat_thread = threading.Thread(target=_post_upload_heartbeat, args=(heartbeat_info,))
+        heartbeat_thread.start()
+        for file_path in self.file_names:
+            geometry._web._upload_file(
+                remote_file_name=os.path.basename(file_path),
+                file_name=file_path,
+                progress_callback=progress_callback,
             )
-            remote_file_names.append(remote_file_name)
-        submitted_geometry._complete_upload(remote_file_names)
-        log.info(f"Geometry successfully submitted: {submitted_geometry.short_description()}")
-        return submitted_geometry
+        heartbeat_info["stop"] = True
+        heartbeat_thread.join()
+        ##:: kick off pipeline
+        RestApi(GeometryInterface.endpoint, id=geometry._web.id).patch(
+            {"action": "Success"}, method="files"
+        )
+        log.info("Geometry successfully submitted.")
+        return geometry
+
+
+class Geometry(AssetBase):
+    """
+    Geometry component for workbench (simulation V2)
+    """
+
+    _interface = GeometryInterface
+    _info_type_class = GeometryMeta
+    _draft_class = GeometryDraft
+    _web: Flow360Resource = None
+
+    @classmethod
+    def from_file(
+        cls,
+        file_names: Union[List[str], str],
+        name: str = None,
+        tags: List[str] = None,
+        length_unit: LengthUnitType = "m",
+    ) -> GeometryDraft:
+        # For type hint only but proper fix is to fully abstract the Draft class too.
+        return super().from_file(file_names, name, tags, length_unit)
+
+    def _retrieve_metadata(self):
+        # get the metadata when initializing the object (blocking)
+        # My next PR
+        pass
+
+    def generate_surface_mesh(self, params: SimulationParams, async_mode: bool = True):
+        """generate surface mesh from the geometry"""
+        return self.run(params, SurfaceMesh, async_mode)
+
+    def generate_volume_mesh(self, params: SimulationParams, async_mode: bool = True):
+        """generate volume mesh from the geometry"""
+        return self.run(params, VolumeMesh, async_mode)
+
+    def run_case(self, params: SimulationParams, async_mode: bool = True):
+        """run case from the geometry"""
+        return self.run(params, Case, async_mode)
```

Comment:
I think these functions should be part of "web" interface and take in Geometry as an argument

Reply:
I think it makes more sense to put them under AssetBase. Web should care more about server communication details (upload/download post patch get etc) and we should not couple the web part with the application-specific details.